### ISSUE TITLE
(9/26) Implement theming

### ISSUE THREAD

**Issue by johngruen** [mozilla/protocol#447]
In order to manage application of Protocol as it applies to Mozilla and Firefox design systems we will need to create a set of theming conventions that cascade down from a top level declarations.

## Requirements
* typographic sizing, color, and font classes should be separate. This separation will allow for down-the-road flexibility for creating different colorways, type scales, and font-usage patterns.
* in order to close this issue we should implement the following:

1. a class that implements the new firefox font stack and default document color scheme w/in a document
2. a class that implements the mozilla font stack and default document color scheme w/in a document
3. a class that implements the firefox type scale w/in a document
4. a class that implements the mozilla type scale w/in a document

## Recommendations

right now, there's an `mzp-t-firefox` class used to set Open Sans...i think we need to be a little more granular about the naming system here, something like this might work: 
```
<body class="mzp-t-brand-firefox mzp-t-usage-presentation">
```

In this version any color and font stack selectors of can be descendants of the `mzp-t-brand-firefox` class, and all scaling selectors can be descendants of of the `mzp-t-usage-presentation` class.

The reason for separating these is because once Protocol is used for app UI, it is very likely that the overall scale will change, so i could easily see a variant here with `mzp-t-usage-application` that keeps the Firefox theme but wants different sizings (and, for example, no `:visited` color on links).

Additionally, this leaves room for another optional base class (`mzp-t-colorway-inverted` or similar) that could override a color theme.

**Comment by johngruen** [comment]
Related: #416

**Comment by stephaniehobson** [comment]
The current set up is that the themes set:
- body font family
- heading font family
- text color
(both themes currently use the same typescale, neither them uses Open Sans anymore)

I don't understand the rational for asking to set typographic sizing, color, and font classes separately. What is the use-case for the mozilla typescale with the firefox font family?

**Comment by johngruen** [comment]
@stephaniehobson the intent here is not to use the mozilla type scale with the firefox font family, it is to separate the typescale from the usage of thefont family...this is a little overhead for the very likely scenario where protocol will be used for application design...for example, in the future we might do something like: `mzp-t-brand-firefox mzp-t-usage-application`

**Comment by ejregithub** [comment]
John to book a meeting Friday a.m. w. Stephanie & Craig

**Comment by craigcook** [comment]
Relevant to themes and different sizes for different fonts, this is some clever work from the USWDS that warrants a look, maybe just for inspiration if not actually adopting the same technique (and also putting the link here before I forget and lose it again): https://designsystem.digital.gov/design-tokens/typesetting/overview/

**Comment by stephaniehobson** [comment]
Here's a proof of concept for how we might handle everything except the light and dark themes (aka colour schemes aka child themes).

https://codepen.io/stephaniehobson/pen/BXwoBK?editors=0100

For people using Protocol they would have to configure global variables for brand and scale before importing components.

```
$brand: 'firefox'; 
$scale: 'marketing';
@import 'protocol/component';
```

When writing components the system uses Sass maps to configure font-family, type scale, and colour pallet. Where possible these value are accessed using a generic named function for example:`color: color-text-display();`.  If it could not be a function it's a mixin like `@include font-size('display', 'md');` (that could be changed to `@include font-size-display-md()` if we want).

**Comment by stephaniehobson** [comment]
Oh, and here are the notes from our meeting: https://docs.google.com/document/d/1DXQ3ZVNyXTGt_869fjeEJpRyX4TyhfYnf_vw7McGdbE/edit?usp=sharing

**Comment by ejregithub** [comment]
@stephaniehobson - are you going to be resuming work on this issue (should I leave it in the Q3 sprint)?

**Comment by stephaniehobson** [comment]
I talked to @craigcook and he's going to be doing implementation work.

**Comment by craigcook** [comment]
Currently I'm achieving themability by assigning all the themable design aspects to variables: change the variables, change the theme. This means turning theme colors into named variables but we haven't really defined a general base palette before so we need to do it now. 

Here are the palette variables I've got so far:
```
background-color
background-color-alt
background-color-inverse
title-text-color
title-text-color-inverse
body-text-color
body-text-color-alt
body-text-color-inverse
link-color
link-color-visited
link-color-hover
link-color-visited-hover
link-color-inverse
link-color-visited-inverse
link-color-hover-inverse
link-color-visited-hover-inverse
```

A few are called "alt" because they're just an alternative color, but maybe there's a better name for those. `background-color-alt` is a light gray but we haven't established any specific purpose for that different color, it's just something other than white when we want some contrast. `body-text-color-alt` is slightly lighter than the default, and is currently in use on form labels, for example. I'm not sure if that lighter text color is only to be used for a particular purpose or if it's just something different. If it's only an alternative color calling it "alt" is appropriate, but if there's a specific meaning attached to that alternative color the name should reflect that meaning.

Also, if we ever want to have multiple alternative colors then "alt" won't work as a name, so maybe it's better better to find a flexible naming convention that gives us room to grow.

The "inverse" colors are for a dark theme (dark background, light text). I like "inverse" better than calling them "light" and "dark" because I worry those would create confusion about when to use which. I like naming things for their purpose and not their appearance. Also wondering if we should have "inverse-alt" background and text colors?

We've talked in the past about doing away with visited link styles so it's possible we could eliminate the visited and visited-hover colors. This seems like a good time to commit to a  decision.

I expect I'll discover more palette colors as I get further into theming all the components. For instance will we need specific button colors, or do buttons just use the text and link colors? There's also [an open issue to define "functional colors"](https://github.com/mozilla/protocol-tokens/issues/53) for alerts and errors. Should these be part of a theme palette as well?
